### DISCUSSION TITLE
Re-organize `filesystem.filters` config options

### DISCUSSION THREAD

**Discussion by cseickel** [nvim-neo-tree/neo-tree.nvim#171] [Ideas]
This is triggered by the work I am doing on #149  and #148, which I am doing together because they are interrelated. As I try to expand on those config options, I realize that it is just confusing because I started out just passing through the options of a Pleanry scan_dir function that no longer really matches what we need.

Here is what I have right now, including the unreleased work in progress of #149 and #148.  **I don't actually like it:**
```lua
require("neo-tree").setup({
  filesystem = {
    filters = {
      show_hidden = false,
      respect_gitignore = true,
      gitignore_source = "git status", -- or "git check-ignored", which may be faster in some repos

      -- The options below are from an unreleased WIP branch:
      exclude_items = {}, -- List of item names to skip, such as:
                          -- { 'node-module', '.DS_Store' }
      show_filtered = { -- Instead of hiding them completely, just mark the node as filtered and let
        exclude = true, -- the components display them differently, such as with a "dim text" highlight group
        gitignore = true,
        hidden = true
      }
    },
  ...
```

### Cleanup Current Option Names
To start off with, is the label `filters` even the best choice? Maybe it should be called `hidden` or `filtered_items`?

Next, `show_hidden` is not quite accurate because it really refers to the *nix concept of hidden, which is to hide dotfiles (starting with a .). I think these should be renamed to `dotfiles`, which also reverses the meaning of boolean value to line up with the gitignore one. I would also change `respect_gitignore` to just `gitignored`, so instead of:

```lua
require("neo-tree").setup({
  filesystem = {
    filters = {
      show_hidden = false,
      respect_gitignore = true,
  ...
```
we have:

```lua
require("neo-tree").setup({
  filesystem = {
    filtered_items = {
      dotfiles = true,
      gitignored = true,
  ...
```

which is the same result as what we have for those settings in the top example but I think the meaning is clearer. 

### Additional Options for Exclude and Show Hidden
I want to discuss these at the same time because those additional options could affect the design of the entire section. To specify specific folders/files to exclude by name, I am proposing either:

```lua
require("neo-tree").setup({
  filesystem = {
    filtered_items = {
      dotfiles = true,
      gitignored = true,
      exclude_by_name = {
        ".DS_Store",
        "thumbs.db",
        "node_modules"
      },
      show_hidden = { -- just alter their display instead of completely removing them from the tree
        dotfiles = true,
        gitignored = true,
        excluded = false
      }
  ...
```

...or maybe the `show_hidden` option can be mixed into the individual items. In the example below, I've added a third state, which is `"never show"`, the meaning of the settings are:

`hidden` = Hide by default, but allow you to toggle the hidden state and show it with a different highlight group or symbol
`marked` = Always show, but with a different highlight group or symbol
`never show` = Always hide it, even if you toggle showing other hidden items, this one will never be shown
EDIT: There really needs to be a fourth item that corresponds to `false`: `"never hide"`

```lua
require("neo-tree").setup({
  filesystem = {
    filtered_items = {
      dotfiles = "hidden", -- or "marked", or "never hide". "never show" probably wouldn't make sense
      gitignored = "marked", -- or "hidden", or "never hide". "never show" probably wouldn't make sense
      exclude_by_name = {
        [".DS_Store"] = "never show", -- or "hidden" or "marked"
        ["thumbs.db"] = "never show", --  or "hidden" or "marked" 
        ["node_modules"] = "marked"  -- or "hidden", "never show" probably wouldn't make sense for this one
      },
    }
  }
})
```

### Toggle Commands
Since I brought up the subject of toggling hidden items above, I also want to mention that I find it annoying it toggle dotfiles and gitgnored with separate commands. Now that we are adding another group for "excluded" items, I think we need to drop the separate commands and just have one single `toggle_all_hidden` command which will show/hide all `filtered_items` at once.

Please let me know what you think and if anyone has any other ideas on this subject.

Thanks!

**Comment by cseickel** [comment]
Another thing that needs to be considered, is that one item may be caught by multiple filters with different settings. For example, given a folder called: `.target` and these filters:

```lua
require("neo-tree").setup({
  filesystem = {
    filtered_items = {
      dotfiles = "hidden", -- or "marked", or "skip". "never show" probably wouldn't make sense
      gitignored = "marked", -- or "hidden", or "skip". "never show" probably wouldn't make sense
      exclude_by_name = {
        [".target"] = "never show",
        [".DS_Store"] = "never show", -- or "hidden" or "marked"
        ["thumbs.db"] = "never show", --  or "hidden" or "marked" 
        ["node_modules"] = "marked"  -- or "hidden", "never show" probably wouldn't make sense for this one
      },
    }
  }
})
```

If `.target` is in the gitignore, then it is caught by every filter, and has the following values:

- "hidden" because it is a `dotfile`
- "marked" because it is `gitignored`
- "never show" because it is marked that way in the `exclude_items` table

So, does the most agressive "never show" override the others, or does the least aggressive "marked" take precedence? Or maybe it goes by strategy and `exclude_by_name` takes precedence over `gitignored` which takes precedence over `dotfiles`? Or the opposite?


Maybe there are too many variations and I should remove the four way value choice? If I go back to true/false and add one more category for `never_show` then it can be:

 ```lua
require("neo-tree").setup({
  filesystem = {
    filtered_items = {
      default_view = "hidden", -- or "marked". This applies to all means of identifying a hidden file.
      dotfiles = true,
      gitignored = true,
      hide_by_name = {
        "node_modules"
      },
      never_show = {
        ".target",
        ".DS_Store",
        "thumbs.db"
      },
    }
  }
})
```

and in that case, `never_show` takes precedence and `.target` is never shown. However, if it was:
 ```lua
require("neo-tree").setup({
  filesystem = {
    filtered_items = {
      default_view = "hidden", -- or "marked". This applies to all means of identifying a hidden file.
      dotfiles = true,
      gitignored = true,
      hide_by_name = {
        "node_modules"
      },
      never_show = {
        ".DS_Store",
        "thumbs.db"
      },
    }
  }
})
```

then `.target` would be hidden by default but can be toggled visible, probably with some markings. It can be marked with both a highlight group for dotfiles and also a `[!!]` or an icon in the `git_status` component.


Sorry for the long winded posts, but the more I think about it the more complicated it gets.

**Comment by nyngwang** [comment]
link my idea on the one of the related thread: https://github.com/nvim-neo-tree/neo-tree.nvim/issues/149#issuecomment-1064102259
